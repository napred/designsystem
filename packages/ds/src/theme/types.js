// @flow

export type FontPalette = {
  default: string,
  [extra: string]: string,
};

export type ThemeSettings = {
  borderRadiuses: Array<string | number>,
  breakpoints: Array<string>,
  colors: { [name: string]: string },
  containerSizes: Array<number | string>,
  fontFamilies?: FontPalette,
  fontSizes: Array<number>,
  gutters: Array<number>,
  importFonts?: Array<string>,
  spacing: Array<number>,
};

export type Theme = {
  get(attributeName: string): any,
};
