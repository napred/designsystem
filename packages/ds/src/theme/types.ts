export interface IFontPalette {
  default: string;
  [extra: string]: string;
}

export interface IThemeSettings {
  borderRadiuses: Array<string | number>;
  breakpoints: string[];
  colors: { [name: string]: string };
  containerSizes: Array<number | string>;
  fontFamilies?: IFontPalette;
  fontSizes: number[];
  gutters: number[];
  importFonts?: string[];
  spacing: number[];
  [key: string]: any;
}
