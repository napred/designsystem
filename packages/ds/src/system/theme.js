// @flow

type FontPalette = {
  default: string,
  [extra: string]: string,
};

type ThemeSettings = {
  borderRadiuses: Array<string | number>,
  breakpoints: Array<string>,
  colors: { [name: string]: string },
  containerSizes: Array<number | string>,
  fontFamilies?: FontPalette,
  fontSizes: Array<number>,
  gutters: Array<number>,
  importFonts?: Array<string>,
  spacing: Array<number>,
};

export default class Theme {
  borderRadiuses: Array<string | number>;
  breakpoints: Array<string>;
  colors: { [name: string]: string };
  containerSizes: Array<number | string>;
  fontFamilies: FontPalette;
  fontSizes: Array<number>;
  gutters: Array<number>;
  importFonts: Array<string>;
  spacing: Array<number>;

  constructor(settings: ThemeSettings) {
    this.borderRadiuses = settings.borderRadiuses;
    this.breakpoints = settings.breakpoints;
    this.colors = settings.colors;
    this.containerSizes = settings.containerSizes;
    this.fontFamilies = settings.fontFamilies || {
      default: 'inherit', // inherit browser defaults
    };
    this.fontSizes = settings.fontSizes;
    this.gutters = settings.gutters;
    this.importFonts = settings.importFonts || [];
    this.spacing = settings.spacing;
  }

  color(name: string, defaultColor?: string): string {
    return (
      this.colors[name] ||
      (defaultColor ? this.colors[defaultColor] || defaultColor : this.colors.grey)
    );
  }

  containerSize(breakpoint: number): string | number {
    return this.containerSizes[breakpoint] || '100%';
  }

  get(attributeName: string): Array<string | number> | { [key: string]: string | number } {
    // $FlowExpectError
    const attributeValues = this[attributeName];

    if (attributeValues == null) {
      throw new Error(`There is no ${attributeName} defined in theme`);
    }

    return attributeValues;
  }

  gutter(breakpoint: number): number {
    return this.gutters[breakpoint] || 0;
  }

  mediaQueries(): Array<string> {
    return this.breakpoints;
  }

  space(scale: number): number {
    if (typeof this.spacing === 'number') {
      return this.spacing;
    }
    console.log('bla');
    return typeof this.spacing[scale] !== 'undefined'
      ? this.spacing[scale]
      : this.spacing[this.spacing.length - 1];
  }

  fontFamily(variant: string): string {
    return this.fontFamilies[variant] || this.fontFamilies.default;
  }

  fontSize(scale: number): number {
    if (typeof this.fontSizes === 'number') {
      return this.fontSizes;
    }
    return typeof this.fontSizes[scale] !== 'undefined'
      ? this.fontSizes[scale]
      : this.fontSizes[this.fontSizes.length - 1];
  }

  fonts(): string {
    const imports = this.importFonts.join(';');

    if (imports !== '') {
      return `${imports};`;
    }

    return '';
  }
}
