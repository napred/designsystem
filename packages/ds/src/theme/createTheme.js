// @flow

import type { ThemeSettings, Theme } from './types';

export default function createTheme(settings: ThemeSettings): Theme {
  return {
    get: (attributeName: string, defaultValue?: any): any => {
      const attributeValues = settings[attributeName];

      if (attributeValues == null) {
        if (defaultValue == null) {
          throw new Error(`There is no ${attributeName} defined in theme`);
        } else {
          return defaultValue;
        }
      } else if (
        typeof attributeValues === 'object' &&
        !Array.isArray(attributeValues) &&
        defaultValue != null
      ) {
        return attributeValues[defaultValue] || defaultValue;
      }

      return attributeValues;
    },
  };
}
