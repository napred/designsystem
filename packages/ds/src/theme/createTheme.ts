import { ITheme } from '../types';
import { IThemeSettings } from './types';

export default function createTheme(settings: IThemeSettings): ITheme {
  return {
    color(valueOrName, defaultValueOrName) {
      const color = settings.colors[valueOrName];

      if (color == null) {
        if (defaultValueOrName != null) {
          return settings.colors[defaultValueOrName] || defaultValueOrName;
        }

        return valueOrName;
      }

      return color;
    },
    get(attributeName, defaultValue) {
      const attributeValues = settings[attributeName];

      if (attributeValues == null) {
        if (defaultValue == null) {
          throw new Error(`There is no ${attributeName} defined in theme`);
        } else {
          return defaultValue;
        }
      } else if (
        typeof attributeValues === 'object' &&
        !Array.isArray(attributeValues) &&
        defaultValue != null
      ) {
        return attributeValues[defaultValue] || defaultValue;
      }

      return attributeValues;
    },

    getResponsiveValue(attributeName, viewport, defaultValue) {
      const attributeValues = settings[attributeName];

      if (attributeValues == null) {
        if (defaultValue == null) {
          throw new Error(`There is no ${attributeName} defined in theme`);
        } else {
          return defaultValue;
        }
      } else if (
        Array.isArray(attributeValues)
      ) {
        return attributeValues[viewport] || attributeValues[attributeValues.length - 1];
      }

      return attributeValues;
    },
  };
}
