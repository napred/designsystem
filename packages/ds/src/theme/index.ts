export * from './types';
export { default as createTheme } from './createTheme';
