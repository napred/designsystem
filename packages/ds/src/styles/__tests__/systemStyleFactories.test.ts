import theme from '../../defaultTheme';
import {
  createNumericSystemStyle,
  createStringSystemStyle,
  createSystemStyle,
} from '../systemStyleFactories';

describe('styler factories', () => {
  describe('createSystemStyle', () => {
    it('works with simple value (without default value)', () => {
      const style = createSystemStyle<{ test?: any }>('test', 'test', 'spacing', val => val);

      expect(style.apply({ test: 1 }, { theme, viewport: 0 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 0 })).toMatchSnapshot();
    });

    it('works with simple value (with default value)', () => {
      const style = createSystemStyle<{ test?: any }>('test', 'test', 'spacing', val => val, 1);

      expect(style.apply({ test: 1 }, { theme, viewport: 0 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 0 })).toMatchSnapshot();
    });

    it('works with responsive value (without default value)', () => {
      const style = createSystemStyle<{ test?: any }>('test', 'test', 'spacing', val => val);

      expect(style.apply({ test: ['1', '2'] }, { theme, viewport: 0 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 0 })).toMatchSnapshot();

      expect(style.apply({ test: ['1', '2'] }, { theme, viewport: 1 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 1 })).toMatchSnapshot();
    });

    it('works with responsive value (with default simple value)', () => {
      const style = createSystemStyle<{ test?: any }>('test', 'test', 'spacing', val => val, '2');

      expect(style.apply({ test: [0] }, { theme, viewport: 0 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 0 })).toMatchSnapshot();
    });

    it('works with responsive value (with default responsive value)', () => {
      const style = createSystemStyle<{ test?: any }>('test', 'test', 'spacing', val => val, [
        '1',
        '2',
      ]);

      expect(style.apply({ test: ['1'] }, { theme, viewport: 0 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 0 })).toMatchSnapshot();
    });

    it('takes the value from default responsive value if is not specified in prop (with default)', () => {
      const style = createSystemStyle<{ test?: any }>('test', 'test', 'spacing', val => val, [
        '1',
        '2',
      ]);

      expect(style.apply({ test: ['1'] }, { theme, viewport: 1 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 1 })).toMatchSnapshot();
    });

    it('takes last value from responsive value if breakpoint exceeds the range (with default responsive)', () => {
      const style = createSystemStyle<{ test?: any }>('test', 'test', 'spacing', val => val, [
        '1',
        '2',
      ]);

      expect(style.apply({ test: ['1'] }, { theme, viewport: 1 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 1 })).toMatchSnapshot();

      expect(style.apply({ test: ['1'] }, { theme, viewport: 2 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 2 })).toMatchSnapshot();

      expect(style.apply({ test: ['1', '3'] }, { theme, viewport: 2 })).toMatchSnapshot();
    });

    it('takes last value from responsive value if breakpoint exceeds the range (with default simple)', () => {
      const style = createSystemStyle<{ test?: any }>('test', 'test', 'spacing', val => val, '1');

      expect(style.apply({ test: ['1'] }, { theme, viewport: 0 })).toMatchSnapshot();
      expect(style.apply({ test: ['1', '3'] }, { theme, viewport: 0 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 0 })).toMatchSnapshot();

      expect(style.apply({ test: ['1'] }, { theme, viewport: 2 })).toMatchSnapshot();
      expect(style.apply({ test: ['1', '3'] }, { theme, viewport: 2 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 2 })).toMatchSnapshot();
    });
  });

  describe('createSystemNumericStyler', () => {
    it('works correctly (returns unit)', () => {
      const style = createNumericSystemStyle<{ test?: any }>('test', 'test', 'spacing');

      expect(style.apply({ test: 1 }, { theme, viewport: 0 })).toMatchSnapshot();
    });
  });

  describe('createSystemStringStyler', () => {
    it('works correctly (returns string)', () => {
      const style = createStringSystemStyle<{ test?: any }>('test', 'test', 'colors');

      expect(style.apply({ test: 'white' }, { theme, viewport: 0 })).toMatchSnapshot();
    });
  });
});
