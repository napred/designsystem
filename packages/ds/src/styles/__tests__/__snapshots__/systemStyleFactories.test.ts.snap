// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styler factories createSystemNumericStyler works correctly (returns unit) 1`] = `
Object {
  "test": "4px",
}
`;

exports[`styler factories createSystemStringStyler works correctly (returns string) 1`] = `
Object {
  "test": "#FFFFFF",
}
`;

exports[`styler factories createSystemStyle takes last value from responsive value if breakpoint exceeds the range (with default responsive) 1`] = `
Object {
  "test": 8,
}
`;

exports[`styler factories createSystemStyle takes last value from responsive value if breakpoint exceeds the range (with default responsive) 2`] = `
Object {
  "test": 8,
}
`;

exports[`styler factories createSystemStyle takes last value from responsive value if breakpoint exceeds the range (with default responsive) 3`] = `
Object {
  "test": 8,
}
`;

exports[`styler factories createSystemStyle takes last value from responsive value if breakpoint exceeds the range (with default responsive) 4`] = `
Object {
  "test": 8,
}
`;

exports[`styler factories createSystemStyle takes last value from responsive value if breakpoint exceeds the range (with default responsive) 5`] = `
Object {
  "test": 16,
}
`;

exports[`styler factories createSystemStyle takes last value from responsive value if breakpoint exceeds the range (with default simple) 1`] = `
Object {
  "test": 4,
}
`;

exports[`styler factories createSystemStyle takes last value from responsive value if breakpoint exceeds the range (with default simple) 2`] = `
Object {
  "test": 4,
}
`;

exports[`styler factories createSystemStyle takes last value from responsive value if breakpoint exceeds the range (with default simple) 3`] = `
Object {
  "test": 4,
}
`;

exports[`styler factories createSystemStyle takes last value from responsive value if breakpoint exceeds the range (with default simple) 4`] = `
Object {
  "test": 4,
}
`;

exports[`styler factories createSystemStyle takes last value from responsive value if breakpoint exceeds the range (with default simple) 5`] = `
Object {
  "test": 16,
}
`;

exports[`styler factories createSystemStyle takes last value from responsive value if breakpoint exceeds the range (with default simple) 6`] = `
Object {
  "test": 4,
}
`;

exports[`styler factories createSystemStyle takes the value from default responsive value if is not specified in prop (with default) 1`] = `
Object {
  "test": 8,
}
`;

exports[`styler factories createSystemStyle takes the value from default responsive value if is not specified in prop (with default) 2`] = `
Object {
  "test": 8,
}
`;

exports[`styler factories createSystemStyle works with responsive value (with default responsive value) 1`] = `
Object {
  "test": 4,
}
`;

exports[`styler factories createSystemStyle works with responsive value (with default responsive value) 2`] = `
Object {
  "test": 4,
}
`;

exports[`styler factories createSystemStyle works with responsive value (with default simple value) 1`] = `
Object {
  "test": 0,
}
`;

exports[`styler factories createSystemStyle works with responsive value (with default simple value) 2`] = `
Object {
  "test": 8,
}
`;

exports[`styler factories createSystemStyle works with responsive value (without default value) 1`] = `
Object {
  "test": 4,
}
`;

exports[`styler factories createSystemStyle works with responsive value (without default value) 2`] = `Object {}`;

exports[`styler factories createSystemStyle works with responsive value (without default value) 3`] = `
Object {
  "test": 8,
}
`;

exports[`styler factories createSystemStyle works with responsive value (without default value) 4`] = `Object {}`;

exports[`styler factories createSystemStyle works with simple value (with default value) 1`] = `
Object {
  "test": 4,
}
`;

exports[`styler factories createSystemStyle works with simple value (with default value) 2`] = `
Object {
  "test": 4,
}
`;

exports[`styler factories createSystemStyle works with simple value (without default value) 1`] = `
Object {
  "test": 4,
}
`;

exports[`styler factories createSystemStyle works with simple value (without default value) 2`] = `Object {}`;
